src/modules/avformat/producer_avformat.c

```diff
@@ -2701,6 +2710,25 @@ static int producer_get_audio( mlt_frame frame, void **buffer, mlt_audio_format
 		AVPacket pkt;
 		mlt_channel_layout layout = mlt_channel_auto;
 
+
+		// Get the audio cache
+		if ( !self->audio_cache )
+		{
+			init_cache(properties, &self->audio_cache);
+		}
+		if ( self->audio_cache )
+		{
+			mlt_frame original = mlt_cache_get_frame( self->audio_cache, position );
+			if ( original )
+			{
+				mlt_frame_get_audio( original, buffer, format, frequency, channels, samples );
```

Comment:
I notice that this directly passes the buffer from the cache frame (rather than make a copy). If a filter is applied to the frame in an attached filter, will the cached buffer be modified by the filter? I am not sure if that issue is mitigated in some non-obvious way.

If it would be better to copy the data, I recommend to use mlt_audio_copy()
https://github.com/mltframework/mlt/blob/master/src/framework/mlt_audio.h#L57

Reply:
I don't really know. Video frames are copied, so copying the audio also is likely the correct approach.

Reply:
I do not want to create extra work, but I think it would be good to consider an implementation that uses mlt_audio in the cache, and copies the data  into the frame.

There is an example here that shows how to move data from mlt_audio to an mlt_frame instance:
https://github.com/mltframework/mlt/blob/master/src/modules/core/link_timeremap.c#L194
Your example reminded me that mlt_channel_layout is not being passed in that code - I should probably fix that.

I think the general process would be:
1) If cache miss: get audio in the normal way, then use mlt_audio_new() to create an audio instance, copy the audio from the frame to the instance and push into the cache.
2) If cache hit: get the mlt_audio instance from the cache and copy into the frame to be output.

Reply:
You do not need to make a copy. `mlt_cache_put_frame()`, which is what this is using, clones the frame with deep copy. That is documented. He is following the pattern done for image cloning, which is a good thing.